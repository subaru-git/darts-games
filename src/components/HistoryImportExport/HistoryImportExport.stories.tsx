/* eslint-disable import/named */
import React from 'react';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import HistoryImportExport from './HistoryImportExport';

export default {
  title: 'Components/HistoryImportExport',
  component: HistoryImportExport,
} as ComponentMeta<typeof HistoryImportExport>;

const Template: ComponentStory<typeof HistoryImportExport> = (args) => (
  <HistoryImportExport {...args} />
);

export const Default = Template.bind({});
Default.args = {};
