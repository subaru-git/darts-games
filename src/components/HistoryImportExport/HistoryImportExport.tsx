import React, { FC, useRef, useState } from 'react'
import {
  AlertDialog,
  AlertDialogBody,
  AlertDialogContent,
  AlertDialogFooter,
  AlertDialogHeader,
  AlertDialogOverlay,
  Button,
  Grid,
  useDisclosure,
} from '@chakra-ui/react'
import { BiExport, BiImport } from 'react-icons/bi'
import useLocale from '@/hooks/locale'
import { exportGameHistory, importGameHistory } from '@/lib/GameHistoryManager/GameHistory'
import { isGameHistory } from '@/lib/Helper/TypeGuard'

type HistoryImportExportProps = {
  onError: () => void
}

const HistoryImportExport: FC<HistoryImportExportProps> = ({ onError }) => {
  const { isOpen, onOpen, onClose } = useDisclosure()
  const [history, setHistory] = useState<GameResult>({ cricketmarkup: [], eagleseye: [] })
  const inputRef = useRef<HTMLInputElement>(null)
  const cancelRef = useRef<HTMLButtonElement>(null)
  const { t } = useLocale()
  return (
    <>
      <Grid templateColumns='repeat(2, 1fr)' gap={4}>
        <input
          type='file'
          accept='.json'
          ref={inputRef}
          hidden
          onChange={(e) => {
            if (!e.target.files) return
            for (const file of e.target.files) {
              const roader = new FileReader()
              roader.onload = (e) => {
                const history = JSON.parse(e.target?.result as string)
                if (!isGameHistory(history)) {
                  onError()
                  return
                }
                setHistory(history)
                onOpen()
              }
              roader.readAsText(file)
            }
            e.target.value = ''
          }}
        />
        <Button
          variant='outline'
          size={{ base: 'sm', md: 'md' }}
          colorScheme='red'
          onClick={() => {
            inputRef.current?.click()
          }}
          leftIcon={<BiImport />}
        >
          Import
        </Button>
        <Button
          variant='outline'
          size={{ base: 'sm', md: 'md' }}
          colorScheme='teal'
          onClick={() => exportGameHistory()}
          leftIcon={<BiExport />}
        >
          Export
        </Button>
      </Grid>
      <AlertDialog
        isOpen={isOpen}
        leastDestructiveRef={cancelRef}
        onClose={onClose}
        size={{ base: 'xs', md: 'md' }}
      >
        <AlertDialogOverlay>
          <AlertDialogContent>
            <AlertDialogHeader fontSize='lg' fontWeight='bold'>
              Import History
            </AlertDialogHeader>
            <AlertDialogBody>{t.IMPORT_DESCRIPTION}</AlertDialogBody>
            <AlertDialogFooter>
              <Button colorScheme='gray' ml={3} onClick={onClose} ref={cancelRef}>
                Cancel
              </Button>
              <Button
                colorScheme='blue'
                onClick={() => {
                  importGameHistory(history, false)
                  onClose()
                }}
                ml={3}
              >
                {t.IMPORT_ADD}
              </Button>
              <Button
                colorScheme='red'
                onClick={() => {
                  importGameHistory(history, true)
                  onClose()
                }}
                ml={3}
              >
                {t.IMPORT_OVERWRITE}
              </Button>
            </AlertDialogFooter>
          </AlertDialogContent>
        </AlertDialogOverlay>
      </AlertDialog>
    </>
  )
}

export default HistoryImportExport
