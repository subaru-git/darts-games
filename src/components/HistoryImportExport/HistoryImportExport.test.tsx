import { render } from '@testing-library/react';
import HistoryImportExport from './HistoryImportExport';

jest.mock('next/router', () => ({ useRouter: jest.fn().mockReturnValue({ locale: 'en' }) }));

test('should rendering', () => {
  const onErrorMock = jest.fn();
  const { container } = render(<HistoryImportExport onError={onErrorMock} />);
  expect(container).toMatchSnapshot();
});
