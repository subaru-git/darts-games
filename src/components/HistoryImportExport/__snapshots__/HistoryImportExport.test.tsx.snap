// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should rendering 1`] = `
<div>
  <div
    class="css-1utx3w7"
  >
    <input
      accept=".json"
      hidden=""
      type="file"
    />
    <button
      class="chakra-button css-4xx2wk"
      type="button"
    >
      <span
        class="chakra-button__icon css-1wh2kri"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m12 18 4-5h-3V2h-2v11H8z"
          />
          <path
            d="M19 9h-4v2h4v9H5v-9h4V9H5c-1.103 0-2 .897-2 2v9c0 1.103.897 2 2 2h14c1.103 0 2-.897 2-2v-9c0-1.103-.897-2-2-2z"
          />
        </svg>
      </span>
      Import
    </button>
    <button
      class="chakra-button css-4xx2wk"
      type="button"
    >
      <span
        class="chakra-button__icon css-1wh2kri"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11 16h2V7h3l-4-5-4 5h3z"
          />
          <path
            d="M5 22h14c1.103 0 2-.897 2-2v-9c0-1.103-.897-2-2-2h-4v2h4v9H5v-9h4V9H5c-1.103 0-2 .897-2 2v9c0 1.103.897 2 2 2z"
          />
        </svg>
      </span>
      Export
    </button>
  </div>
</div>
`;
